compiler/rustc_mir/src/monomorphize/collector.rs

```diff
@@ -59,9 +59,9 @@
 //!
 //! ### Discovering roots
 //!
-//! The roots of the mono item graph correspond to the non-generic
+//! The roots of the mono item graph correspond to the public non-generic
 //! syntactic items in the source code. We find them by walking the HIR of the
-//! crate, and whenever we hit upon a function, method, or static item, we
+//! crate, and whenever we hit upon a public function, method, or static item, we
```

Comment:
In the eager collection mode (as used by incr comp) all non-generic functions are roots. In the lazy collection mode the entry fn (`fn main(){}`), items marked with `#[rustc_std_internal_symbol]` and `is_reachable_non_generic` functions are roots.

https://github.com/rust-lang/rust/blob/a207871d5c22f89093085da89becbb636a45ef89/compiler/rustc_mir/src/monomorphize/collector.rs#L1106-L1120

Reachable non generics includes functions marked with `#[no_mangle]` and all functions in `reachable_set`.

https://github.com/rust-lang/rust/blob/dbc37a97dcf08984163133cff85af5190af037ca/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L44-L144

The `reachable_set` includes all public items, trait implementations and all functions called by inlinable functions.

https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/compiler/rustc_passes/src/reachable.rs#L406-L436